File: ClassePilha/LINGUAGENS/C/pilhas_refatorada.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <assert.h>

// Enumeração para controle rígido do tipo de dado da pilha
typedef enum {
    TIPO_INT,
    TIPO_FLOAT,
    TIPO_CHAR
} TipoDado;

// Union para permitir armazenamento heterogêneo na estrutura física, porém restrito pela pilha
typedef union {
    int v_int;
    float v_float;
    char v_char;
} Elemento;

// Estrutura principal da Pilha
typedef struct {
    Elemento* dados;
    size_t capacidade;
    size_t quantidade;
    TipoDado tipo;
} Pilha;

// Função para criar a pilha com alocação estática inicial do bloco completo de memória
Pilha* criar_pilha(size_t capacidade, TipoDado tipo) {
    if (capacidade == 0) {
        fprintf(stderr, "Erro: A capacidade deve ser maior que zero.\n");
        exit(EXIT_FAILURE);
    }

    Pilha* p = (Pilha*) malloc(sizeof(Pilha));
    if (!p) {
        fprintf(stderr, "Erro de alocacao de memoria para a estrutura.\n");
        exit(EXIT_FAILURE);
    }

    p->capacidade = capacidade;
    p->quantidade = 0;
    p->tipo = tipo;

    // Alocação estática prévia do bloco contíguo total
    p->dados = (Elemento*) malloc(capacidade * sizeof(Elemento));
    if (!p->dados) {
        fprintf(stderr, "Erro de alocacao de memoria para os dados.\n");
        free(p);
        exit(EXIT_FAILURE);
    }

    return p;
}

void destruir_pilha(Pilha* p) {
    if (p) {
        if (p->dados) free(p->dados);
        free(p);
    }
}

bool pilha_esta_vazia(Pilha* p) {
    return p->quantidade == 0;
}

bool pilha_esta_cheia(Pilha* p) {
    return p->quantidade == p->capacidade;
}

size_t tamanho(Pilha* p) {
    return p->quantidade;
}

bool empilha(Pilha* p, Elemento dado) {
    if (pilha_esta_cheia(p)) {
        fprintf(stderr, "PilhaCheiaErro: A pilha esta cheia.\n");
        return false;
    }

    // Insere no topo e incrementa o ponteiro de quantidade
    p->dados[p->quantidade] = dado;
    p->quantidade++;
    return true;
}

bool desempilha(Pilha* p, Elemento* valor_out) {
    if (pilha_esta_vazia(p)) {
        fprintf(stderr, "PilhaVaziaErro: A pilha esta vazia.\n");
        return false;
    }

    p->quantidade--;
    *valor_out = p->dados[p->quantidade];
    return true;
}

bool troca(Pilha* p) {
    if (p->quantidade < 2) {
        fprintf(stderr, "PilhaVaziaErro: Eh necessario ter pelo menos 2 elementos para trocar.\n");
        return false;
    }

    size_t topo = p->quantidade - 1;
    size_t abaixo = p->quantidade - 2;

    Elemento temp = p->dados[topo];
    p->dados[topo] = p->dados[abaixo];
    p->dados[abaixo] = temp;
    return true;
}

// --- BATERIA DE TESTES AUTOMATIZADOS E ESTRESSE ---
int main() {
    printf("==================================================\n");
    printf("     INICIANDO BATERIA DE TESTES DA PILHA (C)     \n");
    printf("==================================================\n\n");

    // 1. Teste com Inteiros
    printf("[TESTE 1] Operacoes Basicas com Inteiros...\n");
    Pilha* pInt = criar_pilha(3, TIPO_INT);
    assert(pilha_esta_vazia(pInt) == true);

    Elemento e1 = {.v_int = 10};
    Elemento e2 = {.v_int = 20};
    Elemento e3 = {.v_int = 30};

    empilha(pInt, e1);
    empilha(pInt, e2);
    empilha(pInt, e3);

    assert(tamanho(pInt) == 3);
    assert(pilha_esta_cheia(pInt) == true);

    Elemento removido;
    desempilha(pInt, &removido);
    assert(removido.v_int == 30);
    printf("  ✓ Elemento desempilhado: %d\n", removido.v_int);

    // Teste da Troca
    troca(pInt);
    desempilha(pInt, &removido);
    assert(removido.v_int == 10);
    printf("  ✓ Elemento apos troca desempilhado: %d\n", removido.v_int);
    printf("  [PASS] Teste 1 concluido com sucesso!\n\n");

    destruir_pilha(pInt);

    // 2. Cenário de Estresse: Pilha Cheia (Overflow)
    printf("[TESTE 2] Cenario de Estresse: Pilha Cheia (Overflow)...\n");
    Pilha* pCheia = criar_pilha(2, TIPO_INT);
    empilha(pCheia, e1);
    empilha(pCheia, e2);

    bool ok = empilha(pCheia, e3);
    assert(ok == false);
    printf("  ✓ Bloqueio de insercao em pilha cheia confirmado.\n");
    printf("  [PASS] Teste 2 (Overflow) concluido com sucesso!\n\n");

    destruir_pilha(pCheia);

    // 3. Cenário de Estresse: Pilha Vazia (Underflow)
    printf("[TESTE 3] Cenario de Estresse: Pilha Vazia (Underflow)...\n");
    Pilha* pVazia = criar_pilha(3, TIPO_INT);

    ok = desempilha(pVazia, &removido);
    assert(ok == false);

    ok = troca(pVazia);
    assert(ok == false);
    printf("  ✓ Bloqueio de remocao e troca em pilha vazia confirmado.\n");
    printf("  [PASS] Teste 3 (Underflow) concluido com sucesso!\n\n");

    destruir_pilha(pVazia);

    // 4. Teste com Caractere (char)
    printf("[TESTE 4] Operacoes com Caractere (char)...\n");
    Pilha* pChar = criar_pilha(2, TIPO_CHAR);
    Elemento c1 = {.v_char = 'A'};
    Elemento c2 = {.v_char = 'B'};

    empilha(pChar, c1);
    empilha(pChar, c2);

    desempilha(pChar, &removido);
    assert(removido.v_char == 'B');
    printf("  ✓ Caractere desempilhado corretamente: %c\n", removido.v_char);
    printf("  [PASS] Teste 4 (char) concluido com sucesso!\n\n");

    destruir_pilha(pChar);

    printf("==================================================\n");
    printf("   TODOS OS TESTES EM C FORAM EXECUTADOS COM SUCESSO! \n");
    printf("==================================================\n");

    return 0;
}

File: ClassePilha/LINGUAGENS/Javascript/pilhas_refatorada.js
// Exceções Customizadas exigidas pelo enunciado
class PilhaCheiaErro extends Error {
    constructor(message = "PilhaCheiaErro: A pilha está cheia.") {
        super(message);
        this.name = "PilhaCheiaErro";
    }
}

class PilhaVaziaErro extends Error {
    constructor(message = "PilhaVaziaErro: A pilha está vazia.") {
        super(message);
        this.name = "PilhaVaziaErro";
    }
}

class TipoErro extends Error {
    constructor(message = "TipoErro: Tipo de dado incompatível.") {
        super(message);
        this.name = "TipoErro";
    }
}

// Classe Pilha com Alocação Física Contígua via TypedArray
class Pilha {
    constructor(capacidade, tipo) {
        if (capacidade <= 0) {
            throw new Error("A capacidade deve ser maior que zero.");
        }

        this.capacidade = capacidade;
        this.tipo = tipo;
        this.quantidade = 0; // Ponteiro manual de quantidade (topo)

        // Instanciação física do TypedArray com bloco contíguo de memória
        if (tipo === Number || tipo === 'int') {
            this.dados = new Int32Array(capacidade);
        } else if (tipo === 'float') {
            this.dados = new Float64Array(capacidade);
        } else if (tipo === String || tipo === 'char') {
            this.dados = new Uint16Array(capacidade); // Armazena UTF-16 code units
        } else {
            throw new TipoErro("Tipo não suportado. Use 'int', 'float' ou 'char'.");
        }
    }

    empilha(dado) {
        if (this.pilha_esta_cheia()) {
            throw new PilhaCheiaErro();
        }

        // Validação estrita do tipo de dado
        if (this.tipo === String || this.tipo === 'char') {
            if (typeof dado !== 'string' || dado.length !== 1) {
                throw new TipoErro("O dado deve ser um único caractere (string de tamanho 1).");
            }
            this.dados[this.quantidade] = dado.charCodeAt(0);
        } else if (this.tipo === Number || this.tipo === 'int' || this.tipo === 'float') {
            if (typeof dado !== 'number' || isNaN(dado)) {
                throw new TipoErro("O dado deve ser um número válido.");
            }
            this.dados[this.quantidade] = dado;
        } else {
            throw new TipoErro();
        }

        this.quantidade++;
    }

    desempilha() {
        if (this.pilha_esta_vazia()) {
            throw new PilhaVaziaErro();
        }

        this.quantidade--;
        let valor = this.dados[this.quantidade];

        if (this.tipo === String || this.tipo === 'char') {
            return String.fromCharCode(valor);
        }
        return valor;
    }

    pilha_esta_vazia() {
        return this.quantidade === 0;
    }

    pilha_esta_cheia() {
        return this.quantidade === this.capacidade;
    }

    troca() {
        if (this.quantidade < 2) {
            throw new PilhaVaziaErro("PilhaVaziaErro: É necessário ter pelo menos 2 elementos para trocar.");
        }

        let topo = this.quantidade - 1;
        let abaixo = this.quantidade - 2;

        let temp = this.dados[topo];
        this.dados[topo] = this.dados[abaixo];
        this.dados[abaixo] = temp;
    }

    tamanho() {
        return this.quantidade;
    }
}

// --- BATERIA DE TESTES AUTOMATIZADOS E ESTRESSE ---
console.log("==================================================");
console.log("   INICIANDO BATERIA DE TESTES DA PILHA (JS)      ");
console.log("==================================================\n");

// 1. Teste de Operações Básicas (Inteiros)
console.log("[TESTE 1] Operações Básicas com Inteiros...");
const pInt = new Pilha(3, 'int');
console.assert(pInt.pilha_esta_vazia() === true, "Erro no teste de pilha vazia");

pInt.empilha(10);
pInt.empilha(20);
pInt.empilha(30);

console.assert(pInt.tamanho() === 3, "Erro no tamanho");
console.assert(pInt.pilha_esta_cheia() === true, "Erro no teste de pilha cheia");

let removido = pInt.desempilha();
console.assert(removido === 30, "Erro ao desempilhar LIFO");
console.log(`  ✓ Elemento desempilhado: ${removido}`);

pInt.troca();
removido = pInt.desempilha();
console.assert(removido === 10, "Erro após operação troca()");
console.log(`  ✓ Elemento desempilhado após troca(): ${removido}`);
console.log("  [PASS] Teste 1 concluído com sucesso!\n");

// 2. Estresse: Overflow (PilhaCheiaErro)
console.log("[TESTE 2] Cenário de Estresse: Pilha Cheia (Overflow)...");
const pCheia = new Pilha(2, 'int');
pCheia.empilha(100);
pCheia.empilha(200);

try {
    pCheia.empilha(300);
    console.assert(false, "Erro: Não lançou PilhaCheiaErro");
} catch (e) {
    console.assert(e instanceof PilhaCheiaErro, "Exceção incorreta");
    console.log(`  ✓ Exceção capturada com sucesso: ${e.message}`);
    console.log("  [PASS] Teste 2 (Overflow) concluído com sucesso!\n");
}

// 3. Estresse: Underflow (PilhaVaziaErro)
console.log("[TESTE 3] Cenário de Estresse: Pilha Vazia (Underflow)...");
const pVazia = new Pilha(3, 'int');

try {
    pVazia.desempilha();
    console.assert(false, "Erro: Não lançou PilhaVaziaErro ao desempilhar");
} catch (e) {
    console.assert(e instanceof PilhaVaziaErro, "Exceção incorreta");
    console.log(`  ✓ Exceção ao desempilhar capturada: ${e.message}`);
}

try {
    pVazia.troca();
    console.assert(false, "Erro: Não lançou PilhaVaziaErro ao trocar");
} catch (e) {
    console.assert(e instanceof PilhaVaziaErro, "Exceção incorreta");
    console.log(`  ✓ Exceção no método troca() capturada: ${e.message}`);
    console.log("  [PASS] Teste 3 (Underflow) concluído com sucesso!\n");
}

// 4. Estresse: Tipo Incompatível (TipoErro)
console.log("[TESTE 4] Cenário de Estresse: Tipo Incompatível...");
const pTipo = new Pilha(3, 'int');

try {
    pTipo.empilha("TextoInvalido");
    console.assert(false, "Erro: Permitiu inserir string em pilha de int");
} catch (e) {
    console.assert(e instanceof TipoErro, "Exceção incorreta");
    console.log(`  ✓ Exceção TipoErro capturada: ${e.message}`);
    console.log("  [PASS] Teste 4 (TipoIncompativel) concluído com sucesso!\n");
}

// 5. Teste com Caractere (char)
console.log("[TESTE 5] Operações com Caractere (char)...");
const pChar = new Pilha(2, 'char');
pChar.empilha('A');
pChar.empilha('B');

let charRemovido = pChar.desempilha();
console.assert(charRemovido === 'B', "Erro no desempilhar de char");
console.log(`  ✓ Caractere desempilhado do Uint16Array: ${charRemovido}`);
console.log("  [PASS] Teste 5 (char) concluído com sucesso!\n");

console.log("==================================================");
console.log("   TODOS OS TESTES EM JS FORAM EXECUTADOS COM SUCESSO!");
console.log("==================================================");